### PR TITLE
WT-13947 Sub-level error codes rollback and compact workflows

### PR DESCRIPTION
**Pull request by DylanSHLiang**
Implemented sub-level error codes for rollback workflows and added unit and python tests for compact and rollback workflows.

### CODE REVIEW

File: test/catch2/misc_tests/test_compact.cpp [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+        // Set background compaction running to true and make the configuration to an empty string.
+        conn_impl->background_compact.running = true;
+        conn_impl->background_compact.config = "";
+
+        CHECK(__wt_background_compact_signal(session_impl, "background=true") == EINVAL);
+        const char *err_msg_content =
+          "Cannot reconfigure background compaction while it's already running.";
+        CHECK(session_impl->err_info.err == EINVAL);
+        CHECK(session_impl->err_info.sub_level_err == WT_BACKGROUND_COMPACT_ALREADY_RUNNING);
+        CHECK(strcmp(session_impl->err_info.err_msg, err_msg_content) == 0);
```
Comment (zunyiliu) [inline test/catch2/misc_tests/test_compact.cpp:41]: I'd prefer making `check_error` sharable and re-using `check_error()` here.

File: test/catch2/misc_tests/test_rollback.cpp [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+#include "wt_internal.h"
+#include "../wrappers/connection_wrapper.h"
+
+/*
+ * [wt_rollback]: test_rollback.cpp
+ * Tests the error handling for rollback workflows.
+ */
+
+void
+check_error(WT_SESSION_IMPL *session, int error, int sub_level_error, std::string error_msg_content)
```
Comment (zunyiliu) [inline test/catch2/misc_tests/test_rollback.cpp:19]: I'm adding a few comments around this function, but they also discuss the broader structure of the code.
1. I think the naming of `test_compact.cpp`, `test_rollback.cpp` are not quite accurate, they are testing if the err msg is recorded properly, but not testing the actual operation.
2. How many `test_xxx.cpp` you reckon we will have for this the project, if there're many of them I think we'd better create a folder and group all of them, e.g. a folder like `/sub_level_err_msg` with all `test_xxx_err_msg.cpp` files beneath it.
3. If 2 is true, we can have a `utils_xxx.cpp` file containing the general functions we will share across `test_xxx_err_msg` files, and the function `check_error()` can be moved to the util file. You can look at`catch2/live_restore/utils_live_restore.cpp` and `catch2/live_restore/utils_live_restore.h` as an example.

File: test/catch2/misc_tests/test_rollback.cpp [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+{
+    CHECK(session->err_info.err == error);
+    CHECK(session->err_info.sub_level_err == sub_level_error);
+    CHECK(session->err_info.err_msg == error_msg_content);
+}
+
+TEST_CASE("Test functions for rollback workflows", "[rollback]")
```
Comment (zunyiliu) [inline test/catch2/misc_tests/test_rollback.cpp:26]: 1. As mentioned above, I think the description should be reworded as this is not testing rollback itself. The `[rollback]` should also be reworded.
2. As mentioned above, if we decide to group all sub_level_error test together, we can have a tag like [sub_level_error] so we can also test all of them together.

File: test/catch2/misc_tests/test_compact.cpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+#include "../wrappers/connection_wrapper.h"
+#include <thread>
+#include <chrono>
+
+/*
+ * [wt_compact]: test_compact.cpp
+ * Tests the error handling for compact workflows.
+ */
+
+TEST_CASE("Test functions for compaction", "[compact]")
```
Comment (zunyiliu) [inline test/catch2/misc_tests/test_compact.cpp:20]: Please refer the comment in code `TEST_CASE("Test functions for rollback workflows", "[rollback]")` as they are the same thing.

File: test/suite/test_error_info.py [hunk 5]
```diff
@@ -52,6 +52,102 @@ def test_compaction_already_running(self):
         # Expect error code, sub-error code and error message to reflect compaction already running.
         self.check_error(errno.EINVAL, wiredtiger.WT_BACKGROUND_COMPACT_ALREADY_RUNNING, "Cannot reconfigure background compaction while it's already running.")
 
+    def test_cache_overflow(self):
```
Comment (zunyiliu) [inline test/suite/test_error_info.py:55]: I think the function name should be reworded, e.g., to `test_rollback_cache_overflow()`. Or at lease we should add a comment telling we are testing cache overflow sub level error for rollback.

File: test/suite/test_error_info.py [hunk 6]
```diff
@@ -52,6 +52,102 @@ def test_compaction_already_running(self):
...
+
+        # This reason is the default reason for WT_ROLLBACK errors so we need to catch it.
+        self.assertRaisesException(wiredtiger.WiredTigerError, lambda: cursor.update())
+
+        # Expect the get_last_error reason to give us the true reason for the rollback.
+        self.check_error(wiredtiger.WT_ROLLBACK, wiredtiger.WT_CACHE_OVERFLOW, "Cache capacity has overflown")
+
+        self.ignoreStdoutPatternIfExists("transaction rolled back because of cache overflow")
+
+    def test_write_conflict(self):
```
Comment (zunyiliu) [inline test/suite/test_error_info.py:87]: Same as above.

File: test/suite/test_error_info.py [hunk 7]
```diff
@@ -52,6 +52,102 @@ def test_compaction_already_running(self):
...
+
+        # Catch the default reason for WT_ROLLBACK errors.
+        self.assertRaisesException(wiredtiger.WiredTigerError, lambda: cursor2.update())
+
+        # Expect the get_last_error reason to give us the true reason for the rollback.
+        # The error will be set in the second session.
+        self.session = session2
+        self.check_error(wiredtiger.WT_ROLLBACK, wiredtiger.WT_WRITE_CONFLICT, "Write conflict between concurrent operations")
+
+    def test_oldest_for_eviction(self):
```
Comment (zunyiliu) [inline test/suite/test_error_info.py:122]: Same as above.

**Review by zunyiliu — CHANGES_REQUESTED** [review]

File: test/catch2/misc_tests/test_compact.cpp [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+    WT_CONNECTION_IMPL *conn_impl = (WT_CONNECTION_IMPL *)conn;
+
+    SECTION("Test __wt_background_compact_signal")
+    {
+        // Set background compaction running to true and make the configuration to an empty string.
+        conn_impl->background_compact.running = true;
+        conn_impl->background_compact.config = "";
+
+        CHECK(__wt_background_compact_signal(session_impl, "background=true") == EINVAL);
+        const char *err_msg_content =
```
[the remark this answers is not in the document]
Comment (DylanSHLiang) [reply test/catch2/misc_tests/test_compact.cpp]: Changed in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/d79d3cd7adab61d7aaea852b41b5406c8d130a3e#diff-434e87a3deeed0a63f4ad36f9eed39e4de68f727dcc2084348a80482cad2d8b8), I just removed the variable and put the message in directly.

File: test/catch2/misc_tests/test_compact.cpp (hunk 1 above)
Comment (DylanSHLiang) [reply test/catch2/misc_tests/test_compact.cpp]: Fixed in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/d79d3cd7adab61d7aaea852b41b5406c8d130a3e#diff-434e87a3deeed0a63f4ad36f9eed39e4de68f727dcc2084348a80482cad2d8b8)

File: test/catch2/misc_tests/test_rollback.cpp (hunk 2 above)
Comment (DylanSHLiang) [reply test/catch2/misc_tests/test_rollback.cpp]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/d79d3cd7adab61d7aaea852b41b5406c8d130a3e#diff-434e87a3deeed0a63f4ad36f9eed39e4de68f727dcc2084348a80482cad2d8b8)

File: test/catch2/misc_tests/test_rollback.cpp (hunk 3 above)
Comment (DylanSHLiang) [reply test/catch2/misc_tests/test_rollback.cpp]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/d79d3cd7adab61d7aaea852b41b5406c8d130a3e#diff-434e87a3deeed0a63f4ad36f9eed39e4de68f727dcc2084348a80482cad2d8b8)

File: test/catch2/misc_tests/test_compact.cpp (hunk 4 above)
Comment (DylanSHLiang) [reply test/catch2/misc_tests/test_compact.cpp]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/d79d3cd7adab61d7aaea852b41b5406c8d130a3e#diff-434e87a3deeed0a63f4ad36f9eed39e4de68f727dcc2084348a80482cad2d8b8)

File: test/suite/test_error_info.py (hunk 5 above)
Comment (DylanSHLiang) [reply test/suite/test_error_info.py]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/d79d3cd7adab61d7aaea852b41b5406c8d130a3e#diff-434e87a3deeed0a63f4ad36f9eed39e4de68f727dcc2084348a80482cad2d8b8)

File: test/suite/test_error_info.py (hunk 6 above)
Comment (DylanSHLiang) [reply test/suite/test_error_info.py]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/d79d3cd7adab61d7aaea852b41b5406c8d130a3e#diff-434e87a3deeed0a63f4ad36f9eed39e4de68f727dcc2084348a80482cad2d8b8)

File: test/suite/test_error_info.py (hunk 7 above)
Comment (DylanSHLiang) [reply test/suite/test_error_info.py]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/d79d3cd7adab61d7aaea852b41b5406c8d130a3e#diff-434e87a3deeed0a63f4ad36f9eed39e4de68f727dcc2084348a80482cad2d8b8)

File: test/catch2/sub_level_err/unit/test_compact_err_msg.cpp [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+/*
+ * [wt_compact]: test_compact_err_msg.cpp
+ * Tests the error handling for compact workflows.
+ */
+
+using namespace utils;
+
+TEST_CASE("Test functions for error handling in compaction workflows",
+  "[compact_err_msg],[sub_level_error]")
```
Comment (zunyiliu) [inline test/catch2/sub_level_err/unit/test_compact_err_msg.cpp:24]: I think using the same prefix for the tests make more sense? like `[sub_level_error_compact]`. I just added a comment here, but there're few other places having this.

Comment (DylanSHLiang) [reply test/catch2/sub_level_err/unit/test_compact_err_msg.cpp]: Changed in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/26a85f11fe136d3756ce0048b78f25dcb6a200e5)

**Review by zunyiliu — APPROVED** [review]
LGTM, please merge develop changes and resolve the conflicts before proceeding.

File: src/evict/evict_lru.c [hunk 10]
```diff
@@ -2955,6 +2958,9 @@ __wti_evict_app_assist_worker(WT_SESSION_IMPL *session, bool busy, bool readonly
 done:
     WT_TRACK_OP_END(session);
 
+    if (ret == WT_ROLLBACK && cache_overflow)
```
Comment (jiechenbo) [inline src/evict/evict_lru.c:2961]: What was the reason again behind the need for the cache_overflow variable?

File: src/include/txn_inline.h [hunk 11]
```diff
@@ -1951,6 +1952,10 @@ __txn_modify_block(
 
 err:
     __wt_scr_free(session, &buf);
+
+    if (ret == WT_ROLLBACK && write_conflict)
```
Comment (jiechenbo) [inline src/include/txn_inline.h:1956]: What's the reason here for the write_conflict variable?

File: test/catch2/sub_level_error/api/test_sub_level_error_session_get_last_error.cpp [hunk 12]
```diff
@@ -8,15 +8,16 @@
...
 
 /*
- * [session_get_last_error]: test_session_get_last_error.cpp
+ * [sub_level_error_session_get_last_error]: test_sub_level_error_session_get_last_error.cpp
  * Tests the API for getting verbose information about the last error of the session.
  */
 
 TEST_CASE("Session get last error - test getting verbose info about the last error in the session",
-  "[session_get_last_error]")
+  "[sub_level_error_session_get_last_error],[sub_level_error]")
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/api/test_sub_level_error_session_get_last_error.cpp:20]: Is there a reason to have two categories here? Do we need to rename the file?

File: test/catch2/sub_level_error/unit/test_sub_level_error_compact.cpp [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+    WT_SESSION *session;
+
+    connection_wrapper conn_wrapper = connection_wrapper(".", "create");
+    WT_CONNECTION *conn = conn_wrapper.get_wt_connection();
+    REQUIRE(conn->open_session(conn, NULL, NULL, &session) == 0);
+    WT_SESSION_IMPL *session_impl = (WT_SESSION_IMPL *)session;
+    WT_CONNECTION_IMPL *conn_impl = (WT_CONNECTION_IMPL *)conn;
+    WT_ERROR_INFO *err_info = &(session_impl->err_info);
+
+    SECTION("Test __wt_background_compact_signal")
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_compact.cpp:35]: I would unit test all of the branch conditions within the if statement and test all the cases where the error is happening or not.

File: test/catch2/sub_level_error/utils_sub_level_error.h [hunk 14]
```diff
@@ -0,0 +1,21 @@
...
+#pragma once
+
+#include <catch2/catch.hpp>
+
+#include "wt_internal.h"
+
+namespace utils {
+
+// Check error helpers
+void check_error_info(
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/utils_sub_level_error.h:18]: I see this was added within Alex's PR #11406, do you know why this change is here too?

File: test/suite/test_error_info.py [hunk 15]
```diff
@@ -52,6 +52,102 @@ def test_compaction_already_running(self):
...
+        cursor = self.session.open_cursor(self.table_name1)
+
+        # Start a new transaction and insert a value far too large for cache.
+        self.session.begin_transaction()
+        cursor.set_key("key_a")
+        cursor.set_value("a"*1024*5000)
+        self.assertEqual(0, cursor.update())
+
+        # Let WiredTiger's accounting catch up.
+        time.sleep(2)
```
Comment (jiechenbo) [inline test/suite/test_error_info.py:138]: Why do we sleep here for 2 seconds? Is it necessary?

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 16]
```diff
@@ -0,0 +1,111 @@
...
+    {
+        // Say that we have 1 change to make and set transaction running to true.
+        session_impl->txn->mod_count = 1;
+        F_SET(session_impl, WT_TXN_RUNNING);
+
+        // Set transaction's ID and pinned ID to be equal to the oldest transaction ID.
+        WT_TXN_SHARED *txn_shared = WT_SESSION_TXN_SHARED(session_impl);
+        txn_shared->id = txn_shared->pinned_id = S2C(session)->txn_global.oldest_id;
+
+        CHECK(__wt_txn_is_blocking(session_impl) == WT_ROLLBACK);
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:100]: I would unit test all of the changes that to returns 0 as well, and making sure that the error message is not set afterwards.

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+        REQUIRE(session->create(session, "table:rollback", "key_format=S,value_format=S") == 0);
+        session->open_cursor(session, "table:rollback", NULL, NULL, &cursor);
+        session->begin_transaction(session, NULL);
+        cursor->set_key(cursor, "key");
+        cursor->set_value(cursor, "value");
+        cursor->update(cursor);
+        session->commit_transaction(session, NULL);
+        cursor->close(cursor);
+
+        CHECK(__wti_evict_app_assist_worker(session_impl, false, false, 100) == WT_ROLLBACK);
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:54]: I wonder what is the behaviour of this function? Does it evict the page in the end?

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+        // WT_UPDATE_TOMBSTONE (4).
+        WT_UPDATE *upd;
+        REQUIRE(__wt_upd_alloc(session_impl, NULL, WT_UPDATE_TOMBSTONE, &upd, NULL) == 0);
+
+        // Transaction must be invisible, so we say that the session has a transaction snapshot and
+        // that the transaction ID is greater than the max snap transaction ID.
+        F_SET(session_impl->txn, WT_TXN_HAS_SNAPSHOT);
+        session_impl->txn->snapshot_data.snap_max = 0;
+        upd->txnid = 1;
+        CHECK(__txn_modify_block(session_impl, NULL, upd, NULL));
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:78]: I see the function has multiple pathways to create the write conflict, could you explore all of them and see if it's possible to write unit test cases for them?

File: src/evict/evict_lru.c (hunk 10 above)
Comment (DylanSHLiang) [reply src/evict/evict_lru.c:2961]: The reason was because a rollback error that occurred lower in the stack would bubble up. Because it's also a a rollback error, this makes sure that it doesn't overwrite it.

File: src/include/txn_inline.h (hunk 11 above)
Comment (DylanSHLiang) [reply src/include/txn_inline.h:1956]: Actually I think I missed this, I made it similar to the one at the top to make sure it doesn't overwrite things, but I'm pretty sure no rollback errors would bubble up to this one. I will remove the write_conflict variable.

File: test/catch2/sub_level_error/api/test_sub_level_error_session_get_last_error.cpp (hunk 12 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/api/test_sub_level_error_session_get_last_error.cpp:20]: The reason is because I wanted to make it similar to the live restore tests to make it consistent. The live restore tests all have the prefix "test_live_restore_xxx". The two categories is so we can run these tests individually, or run all the sub_level_error tests.

File: test/catch2/sub_level_error/utils_sub_level_error.h (hunk 14 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/utils_sub_level_error.h:18]: I've moved Alex's check_error_info helper function to this PR because it's a part of refactoring the structure of the tests. I suggested moving the restructuring to a separate PR, but Zunyi said I can just have it in here

File: test/suite/test_error_info.py (hunk 15 above)
Comment (DylanSHLiang) [reply test/suite/test_error_info.py:138]: This was from the test_txn27.py, I moved the same test to this file.

**Comment by DylanSHLiang** [comment]
Unit tests for WT_WRITE_CONFLICT will be completed in a separate PR.

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 18 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:116]: This is will be sectioned out to a separate PR.

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 16 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:180]: Done in this [PR](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/5eb9ce551c8afb39571774ebafcaf04ee0674cc4)

File: test/catch2/sub_level_error/unit/test_sub_level_error_compact.cpp (hunk 13 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_compact.cpp]: Done in this [PR](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/a150f0b51bd7dbe0b10f5dacd64f0492fa90f010)

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 19]
```diff
@@ -0,0 +1,187 @@
...
+        session_impl->txn->snapshot_data.snap_max = 0;
+        upd->txnid = 1;
+        CHECK(__txn_modify_block(session_impl, NULL, upd, NULL) == WT_ROLLBACK);
+        check_error_info(
+          err_info, WT_ROLLBACK, WT_WRITE_CONFLICT, "Write conflict between concurrent operations");
+
+        // Free update.
+        __wt_free(session_impl, upd);
+
+        // Clear lock so the table can be dropped.
```
Comment (zunyiliu) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:123]: same here.

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 20]
```diff
@@ -0,0 +1,187 @@
...
+        FLD_CLR(session_impl->lock_flags, WT_SESSION_LOCKED_HANDLE_LIST);
+        session->drop(session, "table:rollback", NULL);
+    }
+
+    SECTION("Test WT_OLDEST_FOR_EVICTION in __wt_txn_is_blocking - prepared transaction")
+    {
+        // Set transaction as prepared.
+        F_SET(session_impl->txn, WT_TXN_PREPARE);
+        CHECK(__wt_txn_is_blocking(session_impl) == 0);
+        check_error_info(err_info, 0, WT_NONE, "");
```
Comment (zunyiliu) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:133]: Can you explain the purpose of this test, `__wt_txn_is_blocking()` will run successfully when `WT_TXN_PREPARE` is set?

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 21]
```diff
@@ -0,0 +1,187 @@
...
+    }
+
+    SECTION("Test WT_OLDEST_FOR_EVICTION in __wt_txn_is_blocking - rollback can't be handled")
```
Comment (zunyiliu) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:136]: Same for this SECTION, all the `check_error_info` are expecting `WT_NONE`, I think you can add some comments explaining `WT_NONE` is as expected.

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 20 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:133]: This is an early return from the function. I'm testing it to make sure that the WT_ROLLBACK error isn't returned and that the err struct isn't set.

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 21 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:136]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/8bb8732a5e1d0853918b9eb19b9764c53c2bc7ce)

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 19 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/8bb8732a5e1d0853918b9eb19b9764c53c2bc7ce)

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 22]
```diff
@@ -0,0 +1,187 @@
...
+        CHECK(__wti_evict_app_assist_worker(session_impl, false, false, 100) == WT_ROLLBACK);
+        check_error_info(err_info, WT_ROLLBACK, WT_CACHE_OVERFLOW, "Cache capacity has overflown");
+
+        // Drop the table.
+        session->drop(session, "table:rollback", NULL);
+    }
+
+    SECTION("Test WT_WRITE_CONFLICT in __txn_modify_block")
+    {
+        // Create a table and place a lock on it so session can have a set dhandle.
```
[the remark this answers is not in the document]
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/8bb8732a5e1d0853918b9eb19b9764c53c2bc7ce)

File: test/catch2/sub_level_error/utils_sub_level_error.h (hunk 14 above)
Comment (zunyiliu) [reply test/catch2/sub_level_error/utils_sub_level_error.h:18]: Each PR should have a one-one mapping to a ticket, if you'd like to create a restructuring ticket I'm happy to separate the PR.

As for `check_error_info()`, I think it's fine to add this in both PR otherwise the code won't be able to compile. Once one of the PR is merged into develop, the other may see the conflict and will need to merge the change anyway.

File: test/catch2/sub_level_error/unit/test_sub_level_error_compact.cpp [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+
+        // Set new background compaction config string to true.
+        CHECK(__wt_background_compact_signal(session_impl, "background=true") == 0);
+        check_error_info(err_info, 0, WT_NONE, "");
+    }
+
+    SECTION("Test __wt_background_compact_signal - current background_compact configuration")
+    {
+        // Set current background compaction running to true and expect the current background
+        // configuration to match the new configuration.
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_compact.cpp:61]: Maybe "match the already set configuration" is clearer what do you think?

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 24]
```diff
@@ -0,0 +1,186 @@
...
+
+    SECTION("Test WT_CACHE_OVERFLOW in __wti_evict_app_assist_worker - pre-fetch threads")
+    {
+        F_SET(session_impl, WT_SESSION_PREFETCH_THREAD);
+        CHECK(__wti_evict_app_assist_worker(session_impl, false, false, 100) == 0);
+        check_error_info(err_info, 0, WT_NONE, "");
+    }
+
+    SECTION(
+      "Test WT_CACHE_OVERFLOW in __wti_evict_app_assist_worker - eviction server threads aren't "
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:41]: I'm confused with this one, can you provide more details on what setting up entails?

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 25]
```diff
@@ -0,0 +1,186 @@
...
+        CHECK(__wti_evict_app_assist_worker(session_impl, true, false, 99) == 0);
+        check_error_info(err_info, 0, WT_NONE, WT_ERROR_INFO_SUCCESS);
+
+        conn_impl->evict_server_running = true;
+
+        CHECK(__wti_evict_app_assist_worker(session_impl, false, false, 100) == 0);
+        check_error_info(err_info, 0, WT_NONE, WT_ERROR_INFO_SUCCESS);
+    }
+
+    SECTION("Test WT_CACHE_OVERFLOW in __wti_evict_app_assist_worker - different rollback error")
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:56]: What do we mean by different rollback error?

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 26]
```diff
@@ -0,0 +1,186 @@
...
+        // Set updates to 0.
+        session_impl->txn->mod_count = 0;
+        CHECK(__wt_txn_is_blocking(session_impl) == 0);
+        check_error_info(err_info, 0, WT_NONE, "");
+    }
+
+    SECTION("Test WT_OLDEST_FOR_EVICTION in __wt_txn_is_blocking - transaction ID")
+    {
+        // Say that we have 1 update.
+        session_impl->txn->mod_count = 1;
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:164]: I find the comment confusing to the mod_count variable here, does mod_count relate the number of updates?

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 27]
```diff
@@ -0,0 +1,186 @@
...
+
+        // Set transaction's pinned ID to be equal to the oldest transaction ID.
+        WT_TXN_SHARED *txn_shared = WT_SESSION_TXN_SHARED(session_impl);
+        txn_shared->pinned_id = S2C(session)->txn_global.oldest_id;
+        CHECK(__wt_txn_is_blocking(session_impl) == WT_ROLLBACK);
+        check_error_info(err_info, WT_ROLLBACK, WT_OLDEST_FOR_EVICTION,
+          "Transaction has the oldest pinned transaction ID");
+
+        // Set transaction's ID to be equal to the oldest transaction ID.
+        txn_shared->id = S2C(session)->txn_global.oldest_id;
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:178]: Wouldn't the pinned_id set prior to this case already give the same error? Do we need to reset pinned id back?

File: test/catch2/sub_level_error/utils_sub_level_error.h (hunk 14 above)
Comment (jiechenbo) [reply test/catch2/sub_level_error/utils_sub_level_error.h:18]: Yeah i think it's fine

File: test/suite/test_error_info.py [hunk 28]
```diff
@@ -101,6 +199,18 @@ def test_einval_wt_background_compaction_already_running(self):
         self.api_call_with_einval_wt_background_compaction_already_running()
         self.assert_error_equal(errno.EINVAL, wiredtiger.WT_BACKGROUND_COMPACT_ALREADY_RUNNING, "Cannot reconfigure background compaction while it's already running.")
 
+    def test_wt_rollback_wt_cache_overflow(self):
+        self.api_call_with_wt_rollback_wt_cache_overflow()
+        self.assert_error_equal(wiredtiger.WT_ROLLBACK, wiredtiger.WT_CACHE_OVERFLOW, "Cache capacity has overflown")
+
+    def test_wt_rollback_wt_write_conflic(self):
```
Comment (jiechenbo) [inline test/suite/test_error_info.py:206]: typo for `write_conflic`

File: test/suite/test_error_info.py [hunk 29]
```diff
@@ -35,7 +35,6 @@
 # test_error_info.py
 #   Test that the get_last_error() session API returns the last error to occur in the session.
 class test_error_info(compact_util):
```
Comment (jiechenbo) [inline test/suite/test_error_info.py:37]: This python file is getting too large in my opinion. Can we find a way to section this out? Maybe create test_error_info0 and test_error_info1. Move all the rollback code into test_error_info2.

**Review by jiechenbo — COMMENTED** [review]
Very close to LGTM, just a few comments

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 27 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:178]: Yes, I'm doing this to increase coverage because if the shared id to equal to the oldest id, then it will end the if statement early and not check the pinned id.

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 26 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:164]: I think it does, the if statement says is checks the number of updates and uses this variable

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 24 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:41]: I will change the description. I think it only applies toe the first half of the if statement with the eviction server running variable. It also checks if the application is busy and if the cache is less than 100 percent full

File: test/catch2/sub_level_error/unit/test_sub_level_error_compact.cpp (hunk 23 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_compact.cpp]: I agree, changed in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/5442cfc9772343f4fee61afa8b5990b057c010e3)

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 25 above)
Comment (DylanSHLiang) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp]: I mean the error you get from Oldest for eviction because its possible to get that error, so I've added some comments in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/5442cfc9772343f4fee61afa8b5990b057c010e3)

File: test/suite/test_error_info.py (hunk 28 above)
Comment (DylanSHLiang) [reply test/suite/test_error_info.py]: Fixed in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/5442cfc9772343f4fee61afa8b5990b057c010e3)

File: test/suite/test_error_info.py (hunk 29 above)
Comment (DylanSHLiang) [reply test/suite/test_error_info.py]: Done in this [commit](https://github.com/wiredtiger/wiredtiger/pull/11412/commits/5442cfc9772343f4fee61afa8b5990b057c010e3)

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp (hunk 26 above)
Comment (jiechenbo) [reply test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:169]: nitpick: Rather than say that we have 1 modification, maybe "set the transaction to have 1 modification"

File: test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp [hunk 30]
```diff
@@ -38,23 +38,28 @@ TEST_CASE("Test functions for error handling in rollback workflows",
...
-        CHECK(__wti_evict_app_assist_worker(session_impl, false, false, 100) == 0);
+        // The eviction sever is running, but the application is busy and the cache is less than 100
+        // percent full.
+        CHECK(__wti_evict_app_assist_worker(session_impl, true, false, 99) == 0);
         check_error_info(err_info, 0, WT_NONE, WT_ERROR_INFO_SUCCESS);
     }
 
-    SECTION("Test WT_CACHE_OVERFLOW in __wti_evict_app_assist_worker - different rollback error")
+    SECTION(
+      "Test WT_CACHE_OVERFLOW in __wti_evict_app_assist_worker - WT_OLDEST_FOR_EVICTION rollback "
```
Comment (jiechenbo) [inline test/catch2/sub_level_error/unit/test_sub_level_error_rollback.cpp:57]: Maybe its better to do "test conflicting sub-level error codes"?

File: test/suite/test_error_info02.py [hunk 31]
```diff
@@ -0,0 +1,154 @@
...
+        # This reason is the default reason for WT_ROLLBACK errors so we need to catch it.
+        self.assertRaisesException(wiredtiger.WiredTigerError, lambda: cursor.update())
+
+        self.ignoreStdoutPatternIfExists("transaction rolled back because of cache overflow")
+
+    def api_call_with_wt_rollback_wt_write_conflict(self):
+        """
+        Try to write conflicting data with two threads.
+        """
+        # Create a very basic table.
```
Comment (jiechenbo) [inline test/suite/test_error_info02.py:82]: nitpick: Don't think you need the "very" here.

**Review by jiechenbo — APPROVED** [review]
LGTM just a few comments, no need to re-review from me

**Comment by ershov** [comment]
Hey I think, this PR causes other tasks to fail: https://spruce.mongodb.com/task/wiredtiger_windows_catch2_test_patch_cb3446f8a0acca56416aed13d86f0b1f3c411688_6791c4d3b225920007c227f5_25_01_23_04_25_56/logs?execution=0
